### PR TITLE
[Relayminer] refactor: relayminer to use `claim_window_open_offset_blocks` shared param

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Refactors the relayminer to use the new `ClaimWindowOpenOffsetBlocks` shared param to calculate when to send claim creation messages to the network.

## Issue

- #516 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [ ] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [x] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [x] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: x/shared/session.go [hunk 1]
```diff
@@ -68,3 +68,16 @@ func GetSessionGracePeriodEndHeight(sessionEndHeight int64) int64 {
 func IsGracePeriodElapsed(sessionEndHeight, currentHeight int64) bool {
 	return currentHeight > GetSessionGracePeriodEndHeight(sessionEndHeight)
 }
+
+// GetClaimWindowOpenHeight calculates & returns the earliest block height at which
+// claims can be created for the session which includes queryHeight, given the passed
+// shared module parameters.
```
Comment (bryanchriswhite) [inline x/shared/session.go:74]:
```suggestion
// GetClaimWindowOpenHeight returns the block height at which the claim window opens.
// This height is *inclusive*, meaning that if all other offsets were 0, this is the
// earliest height at which a claim could be committed.
```

File: x/proof/types/shared_query_client.go [hunk 2]
```diff
@@ -29,3 +30,20 @@ func (sharedQueryClient *SharedKeeperQueryClient) GetParams(
 	sharedParams := sharedQueryClient.keeper.GetParams(ctx)
 	return &sharedParams, nil
 }
+
+// GetClaimWindowOpenHeight calculates & returns the earliest block height at which
+// claims can be created for the session which includes queryHeight. It fetches the
+// current shared module params in order to facilitate this.
```
Comment (bryanchriswhite) [inline x/proof/types/shared_query_client.go:36]: #PUC

File: pkg/client/interface.go [hunk 3]
```diff
@@ -285,4 +285,8 @@ type SessionQueryClient interface {
 type SharedQueryClient interface {
 	// GetParams queries the chain for the current shared module parameters.
 	GetParams(ctx context.Context) (*sharedtypes.Params, error)
+	// GetClaimWindowOpenHeight calculates & returns the earliest block height at which
+	// claims can be created for the session which includes queryHeight. It queries the
+	// current queries on-chain params in order to facilitate this.
```
Comment (bryanchriswhite) [inline pkg/client/interface.go:290]: #PUC

File: pkg/client/query/sharedquerier.go [hunk 4]
```diff
@@ -53,3 +54,20 @@ func (sq *sharedQuerier) GetParams(ctx context.Context) (*sharedtypes.Params, er
 	}
 	return &res.Params, nil
 }
+
+// GetClaimWindowOpenHeight calculates & returns the earliest block height at which
+// claims can be created for the session which includes queryHeight. It queries the
+// current on-chain shared module params in order to facilitate this.
```
Comment (bryanchriswhite) [inline pkg/client/query/sharedquerier.go:60]: #PUC

**Review by Olshansk — APPROVED** [review]
Minor comments but no blockers.

File: x/shared/session.go [hunk 5]
```diff
@@ -68,3 +68,15 @@ func GetSessionGracePeriodEndHeight(sessionEndHeight int64) int64 {
 func IsGracePeriodElapsed(sessionEndHeight, currentHeight int64) bool {
 	return currentHeight > GetSessionGracePeriodEndHeight(sessionEndHeight)
 }
+
+// GetClaimWindowOpenHeight returns the block height at which the claim window of
+// the session that includes queryHeight opens, given the passed sharedParams.
+func GetClaimWindowOpenHeight(sharedParams *sharedtypes.Params, queryHeight int64) int64 {
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/shared/session.go:74]: @Olshansk what you linked to in [1] is not the same function as this. Godoc's latest version is v0.0.2, published on May 17, before most of #517 and any of #516 was merged. What you're seeing in godoc is what's being replaced by the work in #517 & #516. Once everything is merged and godoc is up to date, this function should live at [2].

[2] https://pkg.go.dev/github.com/pokt-network/poktroll/x/shared#GetSessionEndBlockHeight

File: pkg/client/query/sharedquerier.go [hunk 6]
```diff
@@ -53,3 +54,19 @@ func (sq *sharedQuerier) GetParams(ctx context.Context) (*sharedtypes.Params, er
...
 }
+
+// GetClaimWindowOpenHeight returns the block height at which the claim window of
+// the session that includes queryHeight opens, given the on-chain shared module params.
+//
+// TODO_TECHDEBT(#543): We don't really want to have to query the params for every method call.
+// Once `ModuleParamsClient` is implemented, use its replay observable's `#Last()` method
+// to get the most recently (asynchronously) observed (and cached) value.
+// TODO_BLOCKER(#543): We also don't really want to use the current value of the params.
+// Instead, we should be using the value that the params had for the session given by blockHeight.
```
Comment (bryanchriswhite) [inline pkg/client/query/sharedquerier.go:65]:
```suggestion
// Instead, we should be using the value that the params had for the session which includes queryHeight.
```

**Pull request merged**